R/stat-sina.r

```diff
@@ -0,0 +1,196 @@
+
+      #find the densiest neighbourhood in the current group and compare it
+      #with the global max
+      tmp_max_neighbours <- max(neighbours[[j]])
+
+      if (tmp_max_neighbours > max_neighbours)
+        max_neighbours <- tmp_max_neighbours
+
+      if (method == "density") {
+
+        #per bin sample density
+        densities[[j]] <- stats::density(data$y[keep], adjust = adjust)
+
+        #find the highest density value
+        tmp_max_density <- max(densities[[j]]$y)
+
+        if (tmp_max_density > max_density)
+          max_density <- tmp_max_density
+      }
+    }
+
+    # ------------------------------------------------------------------------ #
+
+    for (j in levels(factor(data$group))) {
+
+      keep <- data$group == j
+
+      if (nrow(data[keep, ]) < 2 ) next
+
+      #confine the samples in a (-0.5, 0.5) area around the class center
+      if (method == "density") {
+        if (max(densities[[j]]$y) > 0.48)
+          global_scaling_factor <- 0.48 / max(densities[[j]]$y)
+        else
+          global_scaling_factor <- 1
+
+      } else {
+        #if the space required to spread the samples in a neighbourhood exceeds
+        #1, create  compress the points
+        if (max(neighbours[[j]]) > 1 / adjust) {
+          global_scaling_factor <- (1 / adjust) / max(neighbours[[j]])
+        } else
+          global_scaling_factor <- 1
+      }
+
+      #scale all neighbourhoods based on their density relative to the
+      #densiest neighbourhood
+      if (scale == TRUE)
+        group_scaling_factor <- max(neighbours[[j]]) / max_neighbours
+      else
+        group_scaling_factor <- 1
+
+      for (i in names(neighbours[[j]])) {
+
+        #examine neighbourhoods with more than 'neighbour_limit' samples
+        if (neighbours[[j]][i] > neighbour_limit){
+          cur_bin <- bins[ as.integer(i) : (as.integer(i) + 1)]
+
+          #find samples in the current bin and translate their X coord.
+          points <- findInterval(data$y[keep], cur_bin) == 1
+
+          #compute the border margin for the current bin
+          if (method == "density")
+            xmax <- mean(densities[[j]]$y[findInterval(densities[[j]]$x,
+                                                       cur_bin) == 1])
+          else
+            xmax <- adjust * neighbours[[j]][i] / 2
+
+          #assign the samples uniformely within the specified range
+          x_translation <- stats::runif(neighbours[[j]][i], - xmax, xmax )
+
+          #scale and store new x coordinates
+          data$x[keep][points] <- data$x[keep][points] +
+            (x_translation * global_scaling_factor * group_scaling_factor)
+        }
+      }
+    }
+
+    data
+  }
+)
+
+bin_y <- function(data, bw) {
+  #get y value range
+  ymin <- min(data)
+  ymax <- max(data)
+
+  #window width
+  window_size <- (ymax - ymin) * (bw + 1e-8)
+
+  bins <- c()
+  for (i in 0:ceiling(1 / bw)) {
+    bins <- c(bins, ymin + i * window_size)
+  }
+
+  bins
+}
```

Comment:
Not really important, but `bin.R` has some binning functions and using them allows for binning according to the bin width or the number of bins.